pkg/loop/prom.go

```diff
@@ -0,0 +1,106 @@
+package loop
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	"github.com/smartcontractkit/chainlink-relay/pkg/logger"
+)
+
+type PromServer struct {
+	port        int
+	srvrDone    chan struct{} // closed when the http server is done
+	srvr        *http.Server
+	tcpListener *net.TCPListener
+	lggr        logger.Logger
+
+	handler http.Handler
+}
+
+type PromServerOpts struct {
+	Handler http.Handler
+}
+
+func NewPromServer(port int, lggr logger.Logger) *PromServer {
+	return PromServerOpts{}.New(port, lggr)
+}
+
+func (o PromServerOpts) New(port int, lggr logger.Logger) *PromServer {
```

Comment:
@krehermann WDYT about dropping the functional options? I couldn't find satisfactory names (old ones weren't qualified enough in the context of this broader package), and there were extra allocations with the default handler. I'm open to other ideas :shrug:

Reply:
Oh the allocations is a separate problem.

Still not sure about the naming for functional opts though. `WithHandler` is too general for this package namespace, and even `WithPromHandler` has the same problem. So then is `WithPromServerHandler` the best option? That would mean that callers have to do:
`loop.NewPromServer(port, lggr, loop.WithPromServerHandler(handler))`
With an extra `loop.WithPromServer` per option, compared to a struct+method.

Reply:
sorry, seems like being tagged isn't reach me clear thru github. i didn't realize you had asked my opinion here. aside: how do you handle that? it seems to be lost in a stream of other review comment in an email chain for me.

i find options creating an instance confusing. it's easier for me to understand for the constructor of the thing to take options.

is the only use case for testing right now? 

what about something like 

// WithHandler uses the given HTTP handler. It must be called before Start to have any effect
func (*PromServer) WithHandler(h) *PromServer

and the usage would be ~
s := NewPromServer(port,logger).WithHandler(h)

Reply:
> sorry, seems like being tagged isn't reach me clear thru github. i didn't realize you had asked my opinion here. aside: how do you handle that? it seems to be lost in a stream of other review comment in an email chain for me.

Mentions come through my email stream :shrug: 

> i find options creating an instance confusing. it's easier for me to understand for the constructor of the thing to take options.

Can you elaborate on this? I find the opposite, because functional options come with a ton of extra baggage and behave like 'spooky action at a distance'. They strike me as particularly odd when they are only setting fields rather than doing something complex or dynamic.

A `(*PromServer).WithHandler` method seems cleaner than options to me, but it still doesn't seem like it's in the right place if it comes with a caveat that you can only call it at certain times. A basic func or method constructor doesn't come with any extra baggage or caveats - it just does exactly what we need and reads very plainly.

Do you find `net.ListenConfig` difficult to read? https://pkg.go.dev/net#ListenConfig